File: Cpp/Src/src/BarycentricTestResult.cpp
#include "BarycentricTestResult.hpp"


glm::vec3 BarycentricTestResult::weight() const noexcept
{
	return glm::vec3(w1, w2, w3);
}

BarycentricTestResult BarycentricTestResult::test(const glm::vec2 a, const glm::vec2 b, const glm::vec2 c, const double x, const double y) noexcept
{
	const std::function<bool(double)> check = [](double value) {
		return (value >= 0) && (value <= 1.0);
	};

	BarycentricTestResult result;

	//double w1 = a.x * (c.y - a.y) + (y - a.y) * (c.x - a.x) - x * (c.y - a.y);
	//w1 = w1 / ((b.y - a.y) * (c.x - a.x) - (b.x - a.x) * (c.y - a.y));

	//double w2 = y - a.y - w1 * (b.y - a.y);
	//w2 = w2 / (c.y - a.y);

	//double w3 = 1 - w1 - w2;

	const glm::vec3 u = glm::cross(glm::vec3(c.x - a.x, b.x - a.x, a.x - x), glm::vec3(c.y - a.y, b.y - a.y, a.y - y));
	const double w1 = 1.0 - (u.x + u.y) / u.z;
	const double w2 = u.y / u.z;
	const double w3 = u.x / u.z;

	result.w1 = w1;
	result.w2 = w2;
	result.w3 = w3;

	if (check(result.w1) && check(result.w2) && check(result.w3))
	{
		result.isInsideTriangle = true;
	}
	else
	{
		result.isInsideTriangle = false;
	}

	return result;
}

File: Rust/libsr/src/barycentric_test_result.rs
use nalgebra::{Vector2, Vector3};

#[derive(Clone, Copy, Debug)]
pub struct BarycentricTestResult {
    pub is_inside_triangle: bool,
    pub w1: f32,
    pub w2: f32,
    pub w3: f32,
}

impl BarycentricTestResult {
    pub fn test(
        a: &Vector2<f32>,
        b: &Vector2<f32>,
        c: &Vector2<f32>,
        x: f32,
        y: f32,
    ) -> BarycentricTestResult {
        let check = |value: f32| (value >= 0.0) && (value <= 1.0);
        let u = Vector3::new(b.x - a.x, c.x - a.x, a.x - x).cross(&Vector3::new(
            b.y - a.y,
            c.y - a.y,
            a.y - y,
        ));
        let u = u / u.z;
        let w1 = 1.0 - (u.x + u.y);
        let w2 = u.x;
        let w3 = u.y;
        BarycentricTestResult {
            is_inside_triangle: check(w1) && check(w2) && check(w3),
            w1,
            w2,
            w3,
        }
    }

    pub fn weight(&self) -> Vector3<f32> {
        Vector3::new(self.w1, self.w2, self.w3)
    }
}
